src/lib/dom-api-shady.html

```diff
@@ -0,0 +1,632 @@
+    _maybeAddInsertionPoint: function(node, parent) {
+      var added;
+      if (node.nodeType === Node.DOCUMENT_FRAGMENT_NODE &&
+        !node.__noContent) {
+        var c$ = dom(node).querySelectorAll(CONTENT);
+        for (var i=0, n, np, na; (i<c$.length) && (n=c$[i]); i++) {
+          np = dom(n).parentNode;
+          // don't allow node's parent to be fragment itself
+          if (np === node) {
+            np = parent;
+          }
+          na = this._maybeAddInsertionPoint(n, np);
+          added = added || na;
+        }
+      } else if (node.localName === CONTENT) {
+        TreeApi.Logical.saveChildNodes(parent);
+        TreeApi.Logical.saveChildNodes(node);
+        added = true;
+      }
+      return added;
+    },
+
+    _tryRemoveUndistributedNode: function(node) {
+      if (this.node.shadyRoot) {
+        var parent = TreeApi.Composed.getParentNode(node);
+        if (parent) {
+          nativeRemoveChild.call(parent, node);
+        }
+        return true;
+      }
+    },
+
+    _updateInsertionPoints: function(host) {
+      var i$ = host.shadyRoot._insertionPoints =
+        dom(host.shadyRoot).querySelectorAll(CONTENT);
+      // ensure <content>'s and their parents have logical dom info.
+      for (var i=0, c; i < i$.length; i++) {
+        c = i$[i];
+        TreeApi.Logical.saveChildNodes(c);
+        TreeApi.Logical.saveChildNodes(dom(c).parentNode);
+      }
+    },
+
+    _parentNeedsDistribution: function(parent) {
+      return parent && parent.shadyRoot && 
+        DomApi.hasInsertionPoint(parent.shadyRoot);
+    },
+
+    _removeNodeFromParent: function(node) {
+      // note: we may need to notify and not have logical info so fallback
+      // to composed parentNode.
+
+      var parent = node.__parentNode || node.parentNode;
+      //var parent = dom(node).parentNode;
+      // TODO(sorvell): hasDomApi doesn't make sense now
+      if (parent && DomApi.hasApi(parent)) {
+        dom(parent).notifyObserver();
+      }
+      this._removeNodeFromHost(node, true);
+    },
+
+    // NOTE: if `ensureComposedRemoval` is true then the node should be
+    // removed from its composed parent.
+    _removeNodeFromHost: function(node, ensureComposedRemoval) {
+      // note that it's possible for both the node's host and its parent
+      // to require distribution... both cases are handled here.
+      var hostNeedsDist;
+      var root;
+      // important that we want to do this only if the node has a logical parent
+      var parent = node.__parentNode;
+      if (parent) {
+        // distribute node's parent iff needed
+        dom(node)._distributeParent();
+        root = this._ownerShadyRootForNode(node);
+        // remove node from root and distribute it iff needed
+        if (root) {
+          root.host._elementRemove(node);
+          hostNeedsDist = this._removeDistributedChildren(root, node);
+        }
+        TreeApi.Logical.recordRemoveChild(node, parent);
```

Comment:
For add, the logical information is recorded in `_addNode`, the caller of `_addNodeToHost`.  But for remove, the logical information is recorded in `_removeNodeFromHost`, so the factoring just seemed a bit unbalanced...